### PR TITLE
Added quota widget

### LINKED ISSUE #4708: Add quota widget to projects
Design:   https://www.figma.com/file/dJQKgUPzIDND36o2Nq8dct/Loodse-UI?node-id=6869%3A65046

**This feature will be available for enterprise users**

This widget will tell the user how much quota they have consumed and how much is available.

**State of progress bars**
Progress bar < 70& = colour is blue
Progress bar >= 70% = colour turns yellow
Progress bar >= 90% = colour turns red

### PR DESCRIPTION
**Pull request by Talha-Jamil-TJ**
**What does this PR do / Why do we need it**:

Adds quota widget to projects for enterprise users

image

image

image

image

**Which issue(s) this PR fixes** :

Fixes #4708

**Release Note**:

```release-note
Added quota widget for projects
```

### CODE REVIEW

File: src/assets/css/_images.scss [hunk 1]
```diff
@@ -366,6 +366,10 @@
 
 .km-icon-quota {
   @include mixins.mask-image('/assets/images/icons/quota.svg', 24px);
+
+  &.i-32 {
```
Comment (KhizerRehan) [inline src/assets/css/_images.scss:370]: Why do we need this?

`.km-icon-quota` doing same thing this `svg`  exposed as class you can use only `.km-icon-quota` IMO.

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    this._getCurrentUser().then(_ => this._subscribeToQuotaDetails());
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  getProgressBarAccent(percentage: number): ThemePalette {
+    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/component.ts:85]: // eslint-disable-next-line @typescript-eslint/no-magic-numbers
remove this define `readonly` threshhold variabes

-  private _warningThresholdValue: number

Same for ALL

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  constructor(
+    private readonly _cdr: ChangeDetectorRef,
+    private readonly _userService: UserService,
+    private readonly _quotaService: QuotaService
+  ) {}
+
+  ngOnInit(): void {
+    this._setIsEnterpriseEdition();
+    this._getCurrentUser().then(_ => this._subscribeToQuotaDetails());
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/component.ts:76]: try to move logic inside `_initSubscriptions` and remove `async/await` don't use it

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+END OF TERMS AND CONDITIONS
+-->
+
+<div class="km-quota-widget"
+     *ngIf="isEnterpriseEdition && quotaDetails">
+  <ng-container *ngTemplateOutlet="showCompactView ? compactView : extendedView"></ng-container>
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/template.html:25]: can we rename it?
```suggestion
  <ng-container *ngTemplateOutlet="showQuotaDetails ? compactView : extendedView"></ng-container>
```

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+</div>
+
+<ng-template #compactView>
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/template.html:28]:
```suggestion
<ng-template #quotaWidget>
```

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                          [value]="percentage ?? 1"
+                          [matTooltip]="percentage ? percentage + '%' : ''"></mat-progress-bar>
+      </km-property>
+    </ng-template>
+  </ng-template>
+
+
+</ng-template>
+
+<ng-template #extendedView>
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/template.html:102]:
```suggestion
<ng-template #quotaWidgetDetails >
```

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+  private _setIsEnterpriseEdition(): void {
+    this.isEnterpriseEdition = this._getIsEnterpriseEdition();
+  }
+
+  private async _getCurrentUser(): Promise<void> {
+    this._user = await lastValueFrom(this._userService.currentUser.pipe(take(1)));
+  }
+
+  private async _subscribeToQuotaDetails(): Promise<void> {
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/component.ts:102]: remove `async/await` please don't use it.

Moe

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  private readonly _unsubscribe = new Subject<void>();
+  private readonly _debounce = 500;
+  private readonly _showDetails$ = new BehaviorSubject(false);
+  private readonly _getIsEnterpriseEdition = isEnterpriseEdition;
+  private readonly _getPercentage = getPercentage;
+  private _user: Member;
+
+  @Input() projectId: string;
+  @Input() showCompactView = true;
+  @Input() showInCard = true;
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/component.ts:47]:
```suggestion
  @Input() showAsCard = true;
```

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                 let-percentage="percentage">
+      <km-property class="km-quota-item"
+                   fxLayout="column"
+                   fxLayoutAlign="center center">
+        <div key>{{ label }}</div>
+        <mat-progress-bar value
+                          class="property-usage-bar"
+                          mode="determinate"
+                          [color]="getProgressBarAccent(percentage)"
+                          [value]="percentage ?? 1"
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/template.html:93]: why default value `1`? make it `0` i guess?

File: src/app/node-data/component.ts [hunk 10]
```diff
@@ -103,12 +105,14 @@ export class NodeDataComponent extends BaseFormValidator implements OnInit, OnDe
     private readonly _datacenterService: DatacenterService,
     private readonly _nodeDataService: NodeDataService,
     private readonly _settingsService: SettingsService,
+    private readonly _params: ParamsService,
```
Comment (KhizerRehan) [inline src/app/node-data/component.ts:108]:
```suggestion
    private readonly _paramsService: ParamsService,
```

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+  @Input() showIcon = true;
+  @Input() showCompactViewDetailsOnHover = true;
```
Comment (KhizerRehan) [inline src/app/dynamic/enterprise/quotas/quota-widget/component.ts:49]:
```suggestion
  @Input() showDetailsOnHover = true;
```

**Review by KhizerRehan — CHANGES_REQUESTED** [review]
Feedback added

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html (hunk 4 above)
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/quota-widget/template.html:25]: `quota details` refer to the view we have after hovering over the widget, thats why I've named this `compact view`

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html (hunk 9 above)
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/quota-widget/template.html:93]: default value is `1` because when it is `0` the progress bar doesn't show

File: src/app/node-data/component.ts (hunk 10 above)
Comment (Talha-Jamil-TJ) [reply src/app/node-data/component.ts:108]: this is how it's being used throughout the app 
image

File: src/app/dynamic/enterprise/quotas/quota-widget/component.ts (hunk 11 above)
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/quota-widget/component.ts:49]: `showDetailsOnHover` is not specific enough, as the `details view` won't show on hover if the widget is not in `compact view`

File: src/assets/css/_images.scss (hunk 1 above)
Comment (KhizerRehan) [reply src/assets/css/_images.scss:370]: okay fine

File: src/app/node-data/component.ts (hunk 10 above)
Comment (KhizerRehan) [reply src/app/node-data/component.ts:108]: hmm, ideally we should append but either u can skip or change .it

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html (hunk 9 above)
Comment (KhizerRehan) [reply src/app/dynamic/enterprise/quotas/quota-widget/template.html:93]: To add clarification further this is edge case when Quota is ser but `percentage` is `0` so then it make sense to have 1 to at-least show progress bar 👍🏽

**Comment by Waseem826** [comment]
/hold
I would like to test this a bit but if its urgent to merge then please remove this hold.

**Comment by Waseem826** [comment]
/unhold
Please merge this if its ready. I'll test later

File: src/app/shared/utils/common.ts [hunk 12]
```diff
@@ -93,3 +93,8 @@ export function verifyJSON(data: string): boolean {
     return false;
   }
 }
+
+// eslint-disable-next-line @typescript-eslint/no-magic-numbers
```
Comment (ahmedwaleedmalik) [inline src/app/shared/utils/common.ts:97]: Remove this and handle this properly.

Comment (Talha-Jamil-TJ) [reply src/app/shared/utils/common.ts:97]: after storing the default value in `maxUsageDefault` const, IDE is not showing the actual number, but the variable's name,

now: 
image

before:
image

**Comment by Waseem826** [comment]
We should also add this new widget on Project Overview page. There's a separate [ticket](https://github.com/kubermatic/dashboard/issues/4468) for it but we can cover that in this PR.

File: src/app/dynamic/enterprise/quota-widget/style.scss [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    padding-top: 9px;
+
+    .property-usage-bar {
+      @include mixins.size(60px, 5px, true);
+
+      border-radius: variables.$border-radius;
+    }
+
+    [value] {
+      margin: 0.6em 0;
```
Comment (Waseem826) [inline src/app/dynamic/enterprise/quota-widget/style.scss:48]: We're not using relative units anywhere in the app so please change this to `px`. This comment also applies to other `em` usages in this file.

File: src/app/node-data/module.ts [hunk 14]
```diff
@@ -83,6 +84,13 @@ const components = [
   VMwareCloudDirectorBasicNodeDataComponent,
 ];
 
+const routes: Routes = [
```
Comment (Waseem826) [inline src/app/node-data/module.ts:87]: I think we should add a comment here explaining why this is added.

File: src/app/project/component.ts [hunk 15]
```diff
@@ -494,6 +504,21 @@ export class ProjectComponent implements OnInit, OnChanges, OnDestroy {
       });
   }
 
+  onActivate(component: QuotaWidgetComponent, projectId: string, showIcon: boolean): void {
+    component.projectId = projectId;
+    component.showIcon = showIcon;
```
Comment (Waseem826) [inline src/app/project/component.ts:509]:
```suggestion
    component.showIcon = this.showCards;
```

**Review by Waseem826 — CHANGES_REQUESTED** [review]
Some feedback. PTAL

File: src/app/cluster/list/cluster/component.ts [hunk 16]
```diff
@@ -246,6 +249,10 @@ export class ClusterListComponent implements OnInit, OnChanges, OnDestroy {
       : false;
   }
 
+  onActivate(component: QuotaWidgetComponent, projectId: string) {
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/cluster/list/cluster/component.ts]: this will run in a loop, with different project ids

File: src/app/cluster/list/cluster/component.ts [hunk 17]
```diff
@@ -81,7 +82,8 @@ export class ClusterListComponent implements OnInit, OnChanges, OnDestroy {
   ) {}
 
   ngOnInit(): void {
-    this._selectedProject.id = this._activeRoute.snapshot.paramMap.get(PathParam.ProjectID);
+    this.projectId = this._activeRoute.snapshot.paramMap.get(PathParam.ProjectID);
```
[the remark this answers is not in the document]
Comment (Talha-Jamil-TJ) [reply src/app/cluster/list/cluster/component.ts:86]: `selectedProject` is `private`, should I make it `public`?

**Review by Waseem826 — APPROVED** [review]
/approve
